kratos/utilities/element_size_calculator.cpp

```diff
@@ -190,49 +289,317 @@ double ElementSizeCalculator<2,4>::MinimumElementSizeDerivative(
+
+    // Distance between face centers (xi direction)
+    const double dxi_x = x21 - x03;
+    const double dxi_x_derivative = x21_derivative - x03_derivative;
+
+    const double dxi_y = y21 - y03;
+    const double dxi_y_derivative = y21_derivative - y03_derivative;
+
+    const double h2_xi = dxi_x*dxi_x + dxi_y*dxi_y;
+    const double h2_xi_derivative = 2 * dxi_x * dxi_x_derivative + 2 * dxi_y * dxi_y_derivative;
+
+    // Distance between face centers (eta direction)
+    const double deta_x = x32 - x10;
+    const double deta_x_derivative = x32_derivative - x10_derivative;
+
+    const double deta_y = y32 - y10;
+    const double deta_y_derivative = y32_derivative - y10_derivative;
+
+    const double h2_eta = deta_x*deta_x + deta_y*deta_y;
+    const double h2_eta_derivative = 2 * deta_x * deta_x_derivative + 2 * deta_y * deta_y_derivative;
+
+    const double h2 = h2_xi < h2_eta ? h2_xi : h2_eta;
+    const double h2_derivative = h2_xi < h2_eta ? h2_xi_derivative : h2_eta_derivative;
+
+    return 0.5 * h2_derivative / std::sqrt(h2);
+
+    KRATOS_CATCH("");
+}
+
+
+// Tetrahedra3D4 version.
+template<>
+double ElementSizeCalculator<3,4>::MinimumElementSize(const Geometry<Node<3> >& rGeometry)
+{
+
+    /* Calculate distances between each node and the opposite face */
+    const double x10 = rGeometry[1].X() - rGeometry[0].X();
+    const double y10 = rGeometry[1].Y() - rGeometry[0].Y();
+    const double z10 = rGeometry[1].Z() - rGeometry[0].Z();
+
+    const double x20 = rGeometry[2].X() - rGeometry[0].X();
+    const double y20 = rGeometry[2].Y() - rGeometry[0].Y();
+    const double z20 = rGeometry[2].Z() - rGeometry[0].Z();
+
+    const double x30 = rGeometry[3].X() - rGeometry[0].X();
+    const double y30 = rGeometry[3].Y() - rGeometry[0].Y();
+    const double z30 = rGeometry[3].Z() - rGeometry[0].Z();
+
+    // face 123
+    double nx = (y30-y10)*(z20-z10) - (z30-z10)*(y20-y10);
+    double ny = (z30-z10)*(x20-x10) - (x30-x10)*(z20-z10);
+    double nz = (x30-x10)*(y20-y10) - (y30-y10)*(x20-x10);
+    double Hsq = x10*nx + y10*ny + z10*nz; // scalar product x10*n
+    Hsq *= Hsq / (nx*nx + ny*ny + nz*nz); // H^2 = (x10*n)^2 / ||n||^2
+
+    // face 230
+    nx = y30*z20 - z30*y20;
+    ny = z30*x20 - x30*z20;
+    nz = x30*y20 - y30*x20;
+    double hsq = x10*nx + y10*ny + z10*nz;
+    hsq *= hsq / (nx*nx + ny*ny + nz*nz);
+    Hsq = (hsq < Hsq) ? hsq : Hsq;
+
+    // face 301
+    nx = y10*z30 - z10*y30;
+    ny = z10*x30 - x10*z30;
+    nz = x10*y30 - y10*x30;
+    hsq = x20*nx + y20*ny + z20*nz;
+    hsq *= hsq / (nx*nx + ny*ny + nz*nz);
+    Hsq = (hsq < Hsq) ? hsq : Hsq;
+
+    // face 012
+    nx = y10*z20 - z10*y20;
+    ny = z10*x20 - x10*z20;
+    nz = x10*y20 - y10*x20;
+    hsq = x30*nx + y30*ny + z30*nz;
+    hsq *= hsq / (nx*nx + ny*ny + nz*nz);
+    Hsq = (hsq < Hsq) ? hsq : Hsq;
+    return std::sqrt(Hsq);
+}
```

Comment:
Why all these lines appear as modified? I can't see the changes with old implementation.

Reply:
OK. Now I see. GitHub got crazy because the implementation is copied and pasted. Please avoid copying and pasting code by calling the linear tetrahedron function.